src/token_bucket.rs

```diff
@@ -3,6 +3,45 @@ use std::time::{Duration, Instant};
 
 use crate::error::Error;
 
+/// Implementation of the [token bucket](https://en.wikipedia.org/wiki/Token_bucket)
+/// rate-limiting algorithm.
+///
+/// The algorithm is based on the analogy with a fixed capacity *bucket* to
+/// which abstract *tokens* are added with a constant rate up until the bucket
+/// is full.
+///
+/// To implement rate-limiting, each request entering the system is assigned
+/// a *cost* in terms of the number of tokens, and it tries to consume that
+/// many tokens from the bucket:
+///
+/// * if the bucket has the sufficient number of tokens available, they are *consumed*,
+///   and the request is allowed to go through
+///
+/// * otherwise, the request is rejected. New requests continue to be rejected until
+///   the bucket is refilled
+///
+/// Because tokens are added to the bucket with a constant rate over the specified
+/// interval of time, this effectively limits requests to the system to the same rate.
+///
+/// Bucket capacity and token replenishment rate are specified at bucket creation time.
+///
+/// ```
```

Comment:
nit: We probably should add doc tests to our continues integration.

Reply:
Oh, they are already run by `cargo test` by default: https://github.com/ikalnytskyi/youshallnotpass/actions/runs/4450136646/jobs/7815209991?pr=10

```
   Doc-tests youshallnotpass

running 4 tests
test src/token_bucket.rs - token_bucket::TokenBucket<'a>::consume (line 115) ... ok
test src/token_bucket.rs - token_bucket::TokenBucket (line 28) ... ok
test src/token_bucket.rs - token_bucket::TokenBucket<'a>::new (line 56) ... ok
test src/token_bucket.rs - token_bucket::TokenBucket<'a>::new (line 71) ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.40s
```